### PR TITLE
Road to plots 2.0

### PR DESCRIPTION
**Pull request by isentropic**
## What changed
- No default backend
- Backends are purely extensions with no 'requires' 
- The internals are more organized into modules and moved around, thanks to @BeastyBlacksmith 
- Backend init and backend code is only found in its extensions, this should help with backend interference and annoyances about initialization and precompilation of plots. This should make backends modular and implemented much easier and without hacks into Plots.jl codebase

Example:
```
using Plots; import UnicodePlots
unicodeplots()
plot(1:10)
```
## What works
- GR
- UnicodePlot
- 90% of tests pass
- Precompilation

## What does not work
Help needed here add more items or resolve the ticks.

- [x] GR backend has weird axis issue I never got to solve, unicode works consistently same
- [x] plotly backend
- [ ] pythonplot backend
- [x] pgfplotsx backend
- [x] gaston backend
- [x] `default` not resetting legend args

# Help needed:
- [x] Rebase on the master branch
- [ ] Report breakages
- [x] Fix GR backend via help of git diff, something must've changed in GR code
- [ ] move documentation back into this repo
- [ ] Start updating the docs
- [ ] Propose other breaking changes
- [ ] Update the news.md
- [ ] organize a community poll of what defaults could be tweaked and other wishes

# Decisions to be made
- [ ] Deprecate inspectDR?
- [ ] Move to column major order?
- [ ] Change the defaults

Issues needing attention please add more or checkoff resolved:
https://github.com/JuliaPlots/Plots.jl/issues/4565#issue-1466819814
Copy of that:
Labeled issues:
- [ ] https://github.com/JuliaPlots/Plots.jl/issues/4532
- [ ] https://github.com/JuliaPlots/Plots.jl/issues/4028
- [ ] https://github.com/JuliaPlots/Plots.jl/issues/3001
- [ ] https://github.com/JuliaPlots/Plots.jl/issues/2980
- [ ] https://github.com/JuliaPlots/Plots.jl/issues/2129
- [ ] https://github.com/JuliaPlots/Plots.jl/issues/2047
- [x] https://github.com/JuliaPlots/Plots.jl/issues/4181
- [x] https://github.com/JuliaPlots/Plots.jl/issues/4513
- [x] https://github.com/JuliaPlots/Plots.jl/issues/4631
- [ ] https://github.com/JuliaPlots/Plots.jl/issues/4350
- [x] https://github.com/JuliaPlots/Plots.jl/issues/4567
- [ ] https://github.com/JuliaPlots/Plots.jl/issues/4596

Labeled PRs:
- [x] https://github.com/JuliaPlots/Plots.jl/pull/4326
- [ ] https://github.com/JuliaPlots/Plots.jl/pull/3942
- [ ] https://github.com/JuliaPlots/Plots.jl/pull/3744
- [ ] https://github.com/JuliaPlots/Plots.jl/pull/3732
- [ ] https://github.com/JuliaPlots/Plots.jl/pull/3473
- [ ] https://github.com/JuliaPlots/Plots.jl/pull/3357
- [ ] https://github.com/JuliaPlots/Plots.jl/pull/3351
- [ ] Rework adding legends #3732
---
Closes #4326

### CODE REVIEW

File: ext/PlotsGRExt/gr.jl [hunk 1]
```diff
@@ -751,8 +752,8 @@ function gr_get_ticks_size(ticks, rot)
     w, h
 end
 
-function labelfunc(scale::Symbol, backend::GRBackend)
-    texfunc = labelfunc_tex(scale)
+function Plots.labelfunc(scale::Symbol, backend::GRBackend)
+    texfunc = Plots.labelfunc_tex(scale)
```
Comment (BeastyBlacksmith) [inline ext/PlotsGRExt/gr.jl:756]: In general I'd prefer to not have these module declarations inline. It should be visible from the top of the file which functions belong to a different module by 
- `include`ing functions that get extended
- `using` functions, that are used

File: ext/PlotsGRExt/initialization.jl [hunk 2]
```diff
@@ -0,0 +1,207 @@
...
+# ...
+# Plots.supported_attrs(::GRbackend) -> ::Vector{Symbol}
+# ...
+# Plots.supported_scales(::GRbackend) -> ::Vector{Symbol}
+# -----------------------------------------------------------------------------
+
+Plots.is_marker_supported(::GRBackend, shape::Shape) = true
+
+# From, delete this later
+# https://github.com/JuliaLang/Pkg.jl/pull/3552/files#diff-1af5f877eb4497fc1f22daf47044d0958aa02ab39cc6da8ef052624870d75d28R393
```
Comment (BeastyBlacksmith) [inline ext/PlotsGRExt/initialization.jl:207]: what does this relate to?

File: src/Plots.jl [hunk 3]
```diff
@@ -51,7 +50,6 @@ export
     plotarea,
     KW,
 
-    wrap,
```
Comment (BeastyBlacksmith) [inline src/Plots.jl:54]: I think we should keep that (maybe rename it to something like `protect`?)

File: src/abstract_backend.jl [hunk 4]
```diff
@@ -0,0 +1,331 @@
...
+    @eval @require_backend $(backend_package_name(pkg))
+    nothing
+end
+
+# global definitions `const` and `include`
+function _post_imports(pkg::AbstractBackend)
+    name = backend_package_name(pkg)
+    @eval const $name = Main.$name  # so that the module is available in `Plots`
+    nothing
+end
```
Comment (BeastyBlacksmith) [inline src/abstract_backend.jl:250]: I'm not sure we need these anymore

File: src/abstract_backend.jl [hunk 5]
```diff
@@ -0,0 +1,331 @@
...
+
+################################################################################
+# initialize the backends
+function _initialize_backend(pkg::AbstractBackend)
+    _runtime_init(pkg)
+    nothing
+end
+
+# Not sure why these function are here
+# They belong in something like "alignment.jl"
```
Comment (BeastyBlacksmith) [inline src/abstract_backend.jl:263]: agree!

File: test/test_components.jl [hunk 6]
```diff
@@ -1,9 +1,15 @@
 @testset "Shapes" begin
+    Shape = Plots.Shape
+    brush = Plots.brush
+    get_xs = Plots.Shapes.get_xs
+    get_ys = Plots.Shapes.get_ys
+    vertices = Plots.Shapes.vertices
+    rotate! = Plots.Shapes.rotate!
```
Comment (BeastyBlacksmith) [inline test/test_components.jl:7]: apart from Shape we should still export these

**Review by BeastyBlacksmith — COMMENTED** [review]
Thanks for taking this one step further!

File: ext/PlotsGRExt/initialization.jl (hunk 2 above)
Comment (isentropic) [reply ext/PlotsGRExt/initialization.jl:207]: i resolved this, to be deleted

File: ext/PlotsGRExt/gr.jl (hunk 1 above)
Comment (isentropic) [reply ext/PlotsGRExt/gr.jl:756]: Sure let's keep it this way then

File: test/test_components.jl (hunk 6 above)
Comment (isentropic) [reply test/test_components.jl:7]: do you mean that Plots should be exporting these to public, or Shapes exporting these to Plots?

Comment (BeastyBlacksmith) [reply test/test_components.jl:7]: > do you mean that Plots should be exporting these to public

That, its documented API und should stay that way IMO

Comment (isentropic) [reply test/test_components.jl:7]: true, ill update the export, can't find where was this orginally exported

**Comment by isentropic** [comment]
@BeastyBlacksmith I started the implementation of the PGFPlotsX backend, not sure how to proceed with the initialization stage there where you load up more side libraries.
I feel like before the doing any breaking changes and long-awaited PRs we should make the tests pass now and gradually incorporate breaking changes

**Comment by BeastyBlacksmith** [comment]
Sorry for the mess I didn't mean to revert some of your changes, but I can't quite figure out how to recover at the moment...

> how to proceed with the initialization stage there where you load up more side libraries

To which lines are you referring here?

**Comment by isentropic** [comment]
init function in `ext/PlotsPGFPlotsXExt/initialization.jl`

**Comment by isentropic** [comment]
git can be tricky at times...

**Comment by isentropic** [comment]
GR kinda works now, thanks to @jheinen I could debug the issue, it was again unsuprisignly because of some methods not being overloaded due to new modules

**Comment by isentropic** [comment]
Good news, most of GR is passing now:
some errors in the unitful other testset, but nothing critical
```
Test Summary:           | Pass  Fail  Error  Total   Time
backends                |   70     3      5     78  43.8s
  UnicodePlots          |    9                   9   1.2s
  GR - reference images |   60     3      3     66  42.5s
  PlotlyJS              |    1            2      3   0.1s
  Examples              |                     None   0.0s
```

**Review by isentropic — COMMENTED** [review]
thx for finally fixing the tests, its looking good now. Perhaps now we should reintroduce all the backends and fix the rest of the tests. Once this is done, we could merge it to master of Plots and slowly start introducing breaking changes and FRs that were tagged 2.0 while making sure that tests are still passing. And maybe at some point release it.

**Comment by BeastyBlacksmith** [comment]
I'd say we keep it on the v2 branch for a decent while and retarget the v2-PRs to that branch. There is no need to rush this.
Once the v2 branch is in a decent state we should make an announcement on discourse. I am aiming for a release towards the end of the year.

**Comment by isentropic** [comment]
Yeah you are right, given the unpredictability of the development cycle there is no need to be master, but it could be dev. I remember `]add Plots#dev` should work right? I was suggesting this so that we remove friction from trying 2.0 faster and potentially reporting early bugs.

**Comment by BeastyBlacksmith** [comment]
dev is also fine for me

**Comment by isentropic** [comment]
@BeastyBlacksmith you have any idea why the plots is erroring out only because i added some code in the extension? The extension is not really activated even

**Comment by BeastyBlacksmith** [comment]
Which error exactly aer you referring to?

**Comment by isentropic** [comment]
These two things that happenned after my last commits
https://github.com/JuliaPlots/Plots.jl/actions/runs/7609991770/job/20722386227?pr=4866
https://github.com/JuliaPlots/Plots.jl/actions/runs/7609991770/job/20722387137?pr=4866

I'm trying to get pythonplot to work and suddenly tests started failing withuot me enabling the relevant test lines

**Comment by isentropic** [comment]
im also not understanding what is the deal with python plot and the recent breakage with 1.03 version of python plot

**Comment by BeastyBlacksmith** [comment]
> These two things that happenned after my last commits [JuliaPlots/Plots.jl/actions/runs/7609991770/job/20722386227?pr=4866](https://github.com/JuliaPlots/Plots.jl/actions/runs/7609991770/job/20722386227?pr=4866) [JuliaPlots/Plots.jl/actions/runs/7609991770/job/20722387137?pr=4866](https://github.com/JuliaPlots/Plots.jl/actions/runs/7609991770/job/20722387137?pr=4866)
> 
> I'm trying to get pythonplot to work and suddenly tests started failing withuot me enabling the relevant test lines

That is `using Gtk` that errors here.

> im also not understanding what is the deal with python plot and the recent breakage with 1.03 version of python plot

Isn't that https://github.com/JuliaPy/PythonPlot.jl/pull/26#issuecomment-1556799332 ?

**Comment by BeastyBlacksmith** [comment]
I think we are at a point where it makes sense to merge this and then iterate on. I'll copy the OP to the new PR.

**Pull request merged**